### DISCUSSION TITLE
How to send in one Interface While already Listening on a thread?

### DISCUSSION THREAD

**Discussion by primeKal** [python-smpplib/python-smpplib#138] [Q&A]
So i have implemented the library and works very fine:kudos and thanks. But i am already listening on a thread for PDUs and i am saving them to data base...... so my question is in another py file in another interface i am sending a PDU (sending a messages to some SMSC) , so would making a transmitter bind here over ride my threads transreceiver bind? or a can do both? and also when i implement this use case(sending message while already listening in another thread) , my listener receives the delivery report of the other message i sent(but i only want  to save in coming messages), so how can i differentiate a PDU if its a deliver report or a new message. I think my question in another word is how can i use the already bound client object in different pythin codes or can i simply make a new connection and wont affect my first one ie the listnener
Code to listen 

```python
def getPdu(self, pdu):
  #to save the message to database here 
def bind_client(self):
    logging.basicConfig(level='DEBUG')
    client = smpplib.client.Client("","")
    client.set_message_received_handler(
        lambda pdu: self.getPdu(pdu))
    client.set_message_sent_handler(
        lambda pdu: sys.stdout.write('sent {}\n'.format(pdu.receipted_message_id)))
    client.connect()
    client.bind_transceiver(system_id=username, password=password)
    print("binding to client")
    t = Thread(target=client.listen, args=self)
    t.daemon = True
    t.start()
    while not t.is_alive():
        #to notify user if thread is dead
        pass
    return True
```

To sending code

```python
logging.basicConfig(level='DEBUG')

parts, encoding_flag, msg_type_flag = smpplib.gsm.make_parts(msg)
print(parts)

client = smpplib.client.Client(half_url[0], half_url[1])

client.set_message_sent_handler(
    lambda pdu: print("sent"))

def setdlr():
    print("setting")

client.set_message_received_handler(
    lambda pdu: setdlr())

client.connect()
client.bind_transmitter(system_id=username, password=password)

for part in parts:
    pdu = client.send_message(
        source_addr_ton=smpplib.consts.SMPP_TON_INTL,
        # source_addr_npi=smpplib.consts.SMPP_NPI_ISDN,
        # Make sure it is a byte string, not unicode:
        source_addr=fr,

        dest_addr_ton=smpplib.consts.SMPP_TON_INTL,
        # dest_addr_npi=smpplib.consts.SMPP_NPI_ISDN,
        # Make sure thease two params are byte strings, not unicode:
        destination_addr=des,
        short_message=part,
        data_coding=encoding_flag,
        esm_class=msg_type_flag,
        registered_delivery=True,
    )
    self.save_sms(vals)
    client.unbind()
    client.disconnect()
```

**Comment by eigenein** [comment]
Hi @primeKal,

Sorry for the slow response, we have very little resources recently to maintain that.

Anyway,

> I think my question in another word is how can i use the already bound client object in different pythin codes or can i simply make a new connection and wont affect my first one ie the listnener

The client isn't thread-safe per se, so I'd suggest that you make a separate connection.

> so how can i differentiate a PDU if its a deliver report or a new message

For that you look into `pdu.esm_class`: bit `0x04` is a delivery receipt flag (see 4.7.12 `esm_class` SMPP v5 specs).
